cranelift-codegen/meta/src/isa/x86/legalize.rs

```diff
@@ -446,6 +451,39 @@ pub(crate) fn define(shared: &mut SharedDefinitions, x86_instructions: &Instruct
         );
     }
 
+    // SIMD vany_true; `PTEST x, x` will set the ZF flag to 0 if any bits in x contain a 1;
+    // we then set a 1 in a GPR using `SETNZ` (in CLIF, `trueif ne ...`). Remember that the b1
+    // value saved in the GPR by `SETNZ` will not set the upper bits so `bint.i32` or `bextend`
+    // should be used to zero these upper bits. See the Intel manual (vol. 1, appdx. B) to
+    // correlate `ZF = 0` (from PTEST) to `NZ` (used in SETcc) to `ne` (used by CLIF).
+    let ne = Literal::enumerator_for(&imm.intcc, "ne");
+    for ty in ValueType::all_lane_types().filter(allowed_simd_type) {
+        let vany_true = vany_true.bind(vector(ty, sse_vector_size));
+        narrow.legalize(
+            def!(y = vany_true(x)),
+            vec![def!(a = x86_ptest(x, x)), def!(y = trueif(ne, a))],
+        );
+    }
+
+    // SIMD vall_true
+    let zeroes = constant(vec![0x00; 16]);
+    let eq = Literal::enumerator_for(&imm.intcc, "eq");
+    for ty in ValueType::all_lane_types().filter(allowed_simd_type) {
+        let vall_true = vall_true.bind(vector(ty, sse_vector_size));
+        let lane_type_as_int = LaneType::int_from_bits(ty.lane_bits() as u16);
+        let raw_bitcast_to_int = raw_bitcast.bind(vector(lane_type_as_int, sse_vector_size));
+        narrow.legalize(
+            def!(y = vall_true(x)),
+            vec![
+                def!(a = vconst(zeroes)),
+                def!(b = raw_bitcast_to_int(x)),
```

Comment:
I think it would be fine to branch on x's type to avoid this raw_bitcast_to_int. Trying to minimize the number of IR instructions seems like a valuable goal in general, because it reduces the amount of allocated memory in the whole pipeline, makes each optimization pass run faster etc.